ML Engine AI Job submit

callppatel opened the pull request:
<!-- AUTOCHANGELOG for Downstream PRs.

EXTERNAL CONTRIBUTORS: Ignore please - your reviewer will handle.

INTERNAL CONTRIBUTORS AND REVIEWERS: See .ci/RELEASE_NOTES_GUIDE.md
for writing good release notes.

NO CHANGELOG NOTE: Please add "changelog: no-release-note" label to this PR.

Otherwise, fill the template out (replace the heading).
You can add more release notes if you want more than one CHANGELOG entry for
this PR, but make sure not to indent notes and to leave newlines between
code blocks for Markdown's sake.

For Terraform PRs, we use the following "release-note:" headings
    - release-note:enhancement
    - release-note:bug
    - release-note:note
    - release-note:new-resource
    - release-note:new-datasource
    - release-note:deprecation
    - release-note:breaking-change
-->
This is follow-up of issue [https://github.com//issues/2517], I have put very draft version to start with api.yaml. We need terrform to support basic AI job submit. Currently we do with below gcloud command

gcloud ai-platform jobs submit training $JOB_NAME
--job-dir $JOB_DIR
--package-path $TRAINING_PACKAGE_PATH
--module-name $MAIN_TRAINER_MODULE
--region $REGION
--runtime-version=$RUNTIME_VERSION
--python-version=$PYTHON_VERSION
--scale-tier $SCALE_TIER

Release Note Template for Downstream PRs (will be copied)

**Release Note Template for Downstream PRs (will be copied)**

```release-note:new-resource
`google_ml_engine_job`
```

callppatel commented:
@rileykarson - Could you please guide from here. I am able to create simple "trainingInput" job with this code.

modular-magician commented:
Hello!  I am a robot who works on Magic Modules PRs.

I have detected that you are a community contributor, so your PR will be assigned to someone with a commit-bit on this repo for initial review.  They will authorize it to run through our CI pipeline, which will generate downstream PRs.

Thanks for your contribution!  A human will be with you soon.

@danawillow, please review this PR or find an appropriate assignee.

danawillow commented:
Hi @callppatel, I've noticed that you're still adding commits to this. Can you please let me know when it's ready for review? Thanks! You can ignore any warnings that say "This branch is out-of-date with the base branch"

callppatel commented:
@danawillow - This is ready to review with initial draft.

danawillow commented on products/mlengine/ansible.yaml:
```diff
@@ -42,6 +42,8 @@ overrides: !ruby/object:Overrides::ResourceOverrides
 files: !ruby/object:Provider::Config::Files
   resource:
-<%= lines(indent(compile('provider/ansible/resource~compile.yaml'), 4)) -%>
+<%= lines(indent(compile('provider/ansible/resource~compile.yaml'), 4)) -%>
```
nit: can you keep the newline at the end of the file around for files that already had it (and for new files)?

callppatel replied:
done

danawillow commented on products/mlengine/terraform.yaml:
```diff
@@ -33,9 +33,16 @@ overrides: !ruby/object:Overrides::ResourceOverrides
+  Job: !ruby/object:Overrides::Terraform::ResourceOverride
+    properties:
+      jobId: !ruby/object:Overrides::Terraform::PropertyOverride
+        custom_expand: 'templates/terraform/custom_expand/mlengine_job_id.erb'  
```
nit: extra spaces at end of line

callppatel replied:
done

callppatel replied:
I have commented out as of now.

danawillow commented on templates/terraform/custom_expand/mlengine_job_id.erb:
```diff
@@ -0,0 +1,25 @@
+<%# The license inside this block applies to this file.
+	# Copyright 2018 Google Inc.
```
nit:
```suggestion
	# Copyright 2019 Google Inc.
```

callppatel replied:
done

danawillow commented on products/mlengine/api.yaml:
```diff
@@ -235,4 +235,338 @@ objects:
+    properties:
+      - !ruby/object:Api::Type::String
+        name: 'jobId'
+        description: Required. The user-specified id of the job.
```
```suggestion
        description: The user-specified id of the job.
```

danawillow commented on products/mlengine/api.yaml:
```diff
@@ -235,4 +235,338 @@ objects:
+           where both the key and the value are arbitrary Strings that you supply. 
+           An object containing a list of "key": value pairs. Example: { "name": "wrench", "mass": "1.3kg", "count": "3" }.
+      # no etag since it's input: true     
+      ## trainingInput starts here     
```
nit: eh I'm not sure this line is really adding value, since it's pretty easy to see that it starts on the next line.

callppatel replied:
done

danawillow commented on products/mlengine/api.yaml:
```diff
@@ -235,4 +235,338 @@ objects:
+      - !ruby/object:Api::Type::NestedObject
+        name: 'trainingInput'
+        description: Input parameters to create a training job.
+        output: false
```
No need for this- it's false by default

modular-magician commented:
Hi!  I'm the modular magician, I work on Magic Modules.
I see that this PR has already had some downstream PRs generated. Any open downstreams are already updated to your most recent commit, 99d549ed039c2813a70d57e767b39b1c876b56a9. 
## Pull request statuses
No diff detected in Ansible.
No diff detected in Inspec.
## New Pull Requests
I built this PR into one or more new PRs on other repositories, and when those are closed, this PR will also be merged and closed.
depends: https://github.com/terraform-providers/terraform-provider-google-beta/pull/1323 
depends: https://github.com/GoogleCloudPlatform/terraform-google-conversion/pull/247 
depends: https://github.com/terraform-providers/terraform-provider-google/pull/4800

danawillow commented on products/mlengine/api.yaml:
```diff
@@ -235,4 +235,338 @@ objects:
+        properties:
+          - !ruby/object:Api::Type::Enum
+            name: 'scaleTier'
+            description: Specifies the machine types, the number of replicas for workers and parameter servers.
```
Can you also add a description of what each value means (or a link to the documentation of them)?

callppatel replied:
Added the link https://cloud.google.com/ml-engine/reference/rest/v1/projects.jobs#ScaleTier

danawillow commented on products/mlengine/api.yaml:
```diff
@@ -235,4 +235,338 @@ objects:
+            required: true
+          - !ruby/object:Api::Type::String
+            name: 'masterType'
+            description: Specifies the type of virtual machine to use for your training job's master worker  
```
Can you also add the guidance of "You must specify this field when scaleTier is set to CUSTOM."?

danawillow commented on products/mlengine/api.yaml:
```diff
@@ -235,4 +235,338 @@ objects:
+          - !ruby/object:Api::Type::String
+            name: 'masterType'
+            description: Specifies the type of virtual machine to use for your training job's master worker  
+            required: false  
```
false is the default, so this line can be removed (here and elsewhere)

callppatel replied:
Removed

danawillow commented on products/mlengine/api.yaml:
```diff
@@ -235,4 +235,338 @@ objects:
+                  name: 'type'
+                  description: The type of accelerator to use.
+                  values:
+                    - ACCELERATOR_TYPE_UNSPECIFIED
```
I'm not super familiar with the product- is this value one that users might actually set?

callppatel replied:
I have referred below link

https://cloud.google.com/ml-engine/reference/rest/v1/AcceleratorConfig
https://cloud.google.com/ml-engine/reference/rest/v1/AcceleratorType

danawillow commented on products/mlengine/api.yaml:
```diff
@@ -235,4 +235,338 @@ objects:
+                  Otherwise, do not specify this field.
+          - !ruby/object:Api::Type::String
+            name: 'workerType'
+            description: Specifies the type of virtual machine to use for your training job's worker nodes.
```
It looks like there's some further description in the GCP docs, can you add that in? (here and elsewhere)

callppatel replied:
added

danawillow commented on products/mlengine/api.yaml:
```diff
@@ -235,4 +235,338 @@ objects:
+                  Otherwise, do not specify this field.
+          - !ruby/object:Api::Type::String
+            name: 'parameterServerType'
+            description: The configuration for parameter servers
```
This looks like the description for `parameterServerConfig` instead

callppatel replied:
corrected

danawillow commented on products/mlengine/api.yaml:
```diff
@@ -235,4 +235,338 @@ objects:
+      - !ruby/object:Api::Type::NestedObject
+        name: 'trainingOutput'
+        description: The current training job result.
+        output: true
```
This doesn't currently propagate through its nested fields (it probably should), so you'll want to set `output: true` on all of the properties inside of it.

callppatel replied:
ok

danawillow commented on templates/terraform/custom_expand/mlengine_job_id.erb:
```diff
@@ -0,0 +1,25 @@
+	# See the License for the specific language governing permissions and
+	# limitations under the License.
+#%>
+func expand<%= prefix -%><%= titlelize_property(property) -%>(v interface{}, d TerraformResourceData, config *Config) (interface{}, error) {
```
Why are we adding extra information into the job id instead of passing what the user gave us through?

callppatel replied:
We are adding timestamp of the job. Similar to to gcloud command does as per recommendation. Similar to the link https://cloud.google.com/ml-engine/docs/scikit/training-scikit-learn

JOB_NAME=census_training_$(date +"%Y%m%d_%H%M%S")

Job name must be unique each time

callppatel replied:
Will review with team and remove it.

callppatel replied:
removed this

danawillow commented on templates/terraform/examples/ml_job_training.tf.erb:
```diff
@@ -0,0 +1,12 @@
+resource "google_ml_engine_job" "test_job" {
+  job_id          = "ml_job_training"
+  training_input  {
+    job_dir         = "gs://cio-strat-svc-acct-mgmt-238520-mlengine/scikit_learn_job_dir"
```
When I run the test, I get this error:
```
Error 400: Field: job_dir Error: The provided GCS path gs://cio-strat-svc-acct-mgmt-238520-mlengine/scikit_learn_job_dir cannot be written by current user. Please make sure that the bucket exists and you have write access to it.
```

Is it possible to use other TF resources alongside this one to make sure the tests are able to run no matter the environment?

callppatel replied:
yes, I will convert this to parameter

callppatel replied:
Please pass correct data

danawillow commented on templates/terraform/examples/ml_job_training.tf.erb:
```diff
@@ -0,0 +1,12 @@
+resource "google_ml_engine_job" "test_job" {
```
I'd recommend having two tests- a "basic" test where only required fields are set, and a "full" test that attempts to use as much of the resource as possible. Having a test that leaves most optional fields empty allows us to check  for optional fields whether we need to add a default value into api.yaml, mark it as default_from_api in terraform.yaml, or whether neither are necessary (I usually try to do neither until I see a test fail because we got a value back from the server that we didn't send).

callppatel replied:
I will be coming with multiple test case

callppatel replied:
test is added

danawillow commented:
Hi @callppatel, let me know when this is ready for re-review. I'd also like to bump up my top-level comment about your use case because I'd like your opinion: https://github.com/GoogleCloudPlatform/magic-modules/pull/2565#pullrequestreview-310730142

callppatel commented:
@danawillow 

> Thank you for contributing! I've left a bunch of comments, but most of them are small nitpicks. One thing I'd ask overall is whether there's anything in here that can be removed. Right now this resource contains a large number of fields (many of which are output-only), and I want to make sure we don't drown our users in information that they won't actually be using. Since I don't use ML Engine myself, I'll defer to your judgment on whether any of them should be removed. However, I do ask for any that are kept that are settable, that we have a test that sets that field.
> 
> I'm also super curious about your use case for this! Looking through the docs, this specific feature seems pretty imperative- you submit a job because you want it to run right away, and maybe later you'll run another one. In Terraform, trying to create the same resource a second time means tainting it so that it destroys and recreates itself. Can you share a bit on your expected workflow with this resource in Terraform?

This is ready to review, also I need suggestion form Google to proceed further fast pace, so far I have added three test cases. I have included all input fields for ["Traininginput"](https://cloud.google.com/ml-engine/reference/rest/v1/projects.jobs#TrainingInput) only, our use case is only around **TrainingInput**, thats the reason, I have not added anything for [PredictionInput](https://cloud.google.com/ml-engine/reference/rest/v1/projects.jobs#PredictionInput). 
Purpose of Terraform is to submit ML jobs, once job is submitted it will start running and moved to status to completed. There are lot of parameter we pass, I have added all fields of which mentioned in API. 

One of test case I have used which is mention [here](https://cloud.google.com/ml-engine/docs/scikit/training-scikit-learn), currently all these are done thru gcloud commands.  Our organization wants to implement IaC by using Terraform.  I am also new to ML Engine, working as just Magic Module contributor. 

This resource is slightly different from other GCP resource as these will not be destroyed once job status is changed. We want help from your team to suggest to move forward.  As of now, I am able to replicate three test case which was being done thru "gcloud" and now I can submit thru "Terraform" using this.  "Terraform Destroy" doesn't do anything here. So I need suggestion how to handle and what to do.

modular-magician commented:
Hi!  I'm the modular magician, I work on Magic Modules.
I see that this PR has already had some downstream PRs generated. Any open downstreams are already updated to your most recent commit, a4eebdf2b1f034ccf01da4a9cb4cc31ddad9cb93. 
## Pull request statuses
terraform-provider-google-beta already has an open PR.
terraform-google-conversion already has an open PR.
terraform-provider-google already has an open PR.
No diff detected in Ansible.
No diff detected in Inspec.
## New Pull Requests
I didn't open any new pull requests because of this PR.

callppatel commented on products/mlengine/api.yaml:
```diff
@@ -235,4 +235,354 @@ objects:
+  - !ruby/object:Api::Resource
+    name: 'Job'
+    base_url: projects/{{project}}/jobs
+    self_link: projects/{{project}}/jobs/{{name}}
```
Removing    self_link: projects/{{project}}/jobs/{{name}}

callppatel commented on products/mlengine/api.yaml:
```diff
@@ -235,4 +235,354 @@ objects:
+        description: |
+           One or more labels that you can add, to organize your jobs. Each label is a key-value pair, 
+           where both the key and the value are arbitrary Strings that you supply. 
+           An object containing a list of "key": value pairs. Example: { "name": "wrench", "mass": "1.3kg", "count": "3" }.
```
removed

callppatel commented on products/mlengine/api.yaml:
```diff
@@ -235,4 +235,354 @@ objects:
+            name: 'completedTrialCount'
+            description: The number of hyperparameter tuning trials that completed successfully. Only set for hyperparameter tuning jobs.
+            output: true
+          - !ruby/object:Api::Type::NestedObject
```
no, this is string  "completedTrialCount": string
https://cloud.google.com/ml-engine/reference/rest/v1/projects.jobs#TrainingOutput

danawillow replied:
@callppatel, I think you're looking at the definition a few lines up from the one I commented on. I'm specifically talking about the NestedObject 'trials' on the line where I made the comment.

callppatel replied:
Oh, ya. I will fix this.

callppatel commented on products/mlengine/api.yaml:
```diff
@@ -235,4 +235,354 @@ objects:
+                name: 'isTrialStoppedEarly'
+                description:  True if the trial is stopped early.
+                output: true
+              - !ruby/object:Api::Type::NestedObject
```
As per API document, it is boolean 

https://cloud.google.com/ml-engine/reference/rest/v1/projects.jobs#TrainingOutput

danawillow replied:
see other comment for line number clarification

callppatel replied:
got it. I am fixing it.

callppatel replied:
Could you please share dataflow example which needs to handle for "Terraform Destroy", I dont see "dataflow" under "products" folder or magic module. It must be created with custom code.

danawillow replied:
Correct, we don't generate dataflow resources yet. We have the ability to add a full custom delete function, so you can model it off of the dataflow one: https://github.com/GoogleCloudPlatform/magic-modules/blob/master/third_party/terraform/resources/resource_dataflow_job.go

callppatel commented on products/mlengine/api.yaml:
```diff
@@ -235,4 +235,354 @@ objects:
+              - !ruby/object:Api::Type::Integer
+                name: 'pythonVersion'
+                description:  Python version on which the built-in algorithm was trained.          
+              - !ruby/object:Api::Type::Integer
```
ya,I have changed to String

danawillow replied:
same comment here

callppatel replied:
fixing

callppatel commented on templates/terraform/examples/ml_job_training.tf.erb:
```diff
@@ -0,0 +1,13 @@
+    runtime_version = "1.14"
+    scale_tier      = "BASIC"
+  }
+}
```
added new line

callppatel commented:
Hi @danawillow - Your test case are failing because I have not mentioned the exact GCS Bucket name in terraform.yaml file which is just dummy name. I am able to test all three test cases as my end because I am using valid GCS bucket name for my project, which I can't share or provide global access.  Not sure how your test works in such scenario at your end, you might need to set-up. I have fixed or provided input for your review comment.  
Following is still open at my side.
1) Need to discuss Job name convention with my team and let you know, as of now I just follow the GCP docs where the training job is suffix with timestamp. 
2) Regarding destroy, I still need to review dataflow examples.

callppatel commented on products/mlengine/api.yaml:
```diff
@@ -235,4 +235,354 @@ objects:
+                      The Cloud Storage path to the model/ directory where the training job saves the trained model. 
+                      Only set for successful jobs that don't use hyperparameter tuning
+                    output: true
+          - !ruby/object:Api::Type::Integer
```
Shall I change to "String", I dont think Type::Number is a valid..

danawillow replied:
you can use Double

callppatel replied:
fixing

callppatel commented on products/mlengine/api.yaml:
```diff
@@ -235,4 +235,354 @@ objects:
+          - !ruby/object:Api::Type::Boolean
+            name: 'isHyperparameterTuningJob'
+            description: Whether this job is a hyperparameter tuning job.
+          - !ruby/object:Api::Type::String
```
Changed

danawillow replied:
not changed (see other comment about line numbers)

callppatel replied:
fixing

callppatel commented on products/mlengine/api.yaml:
```diff
@@ -235,4 +235,354 @@ objects:
+              - !ruby/object:Api::Type::String
+                name: 'framework'
+                description:  Framework on which the built-in algorithm was trained.
+              - !ruby/object:Api::Type::Integer
```
changed

danawillow replied:
same comment here

callppatel commented on products/mlengine/api.yaml:
```diff
@@ -235,4 +235,354 @@ objects:
+                output: true
+              - !ruby/object:Api::Type::NestedObject
+                name: 'allMetrics'
+                description:  All recorded object metrics for this trial. This field is not currently populated.
```
This is output fields

danawillow replied:
Can you clarify please? I can see that you've marked it as output, but if, as the description says, it isn't populated, then I'm not sure why our users would want this object in their state file. That's the question I'm looking for an answer for.

callppatel replied:
Shall, I exclude this?

danawillow replied:
That's my inclination. In the  initial release of the resource, I'd rather have only the information that we think customers are going to use, since it's easier to add things later than it is to remove them if we discover they're unnecessary or difficult to maintain.

callppatel replied:
Ok, I am removing it. Since it is output, only api.yaml change is needed.

callppatel commented on products/mlengine/api.yaml:
```diff
@@ -235,4 +235,354 @@ objects:
+              - !ruby/object:Api::Type::Integer
+                name: 'runtimeVersion'
+                description:  AI Platform runtime version on which the built-in algorithm was trained         
+              - !ruby/object:Api::Type::Integer
```
resolved

callppatel commented:
> Can you point me to where the objects in the GCS bucket came from (in the GCP docs)? I'll be able to get you a better answer on how to get it working in our test environment after that.

One line is https://cloud.google.com/ml-engine/docs/scikit/training-scikit-learn, I will update with "gs://cloud-samples-data/", hope it will work with this.
Other is
https://cloud.google.com/ml-engine/docs/training-jobs

callppatel commented:
@danawillow - Could you please advise next step on this.

danawillow commented:
You mentioned in https://github.com/GoogleCloudPlatform/magic-modules/pull/2565#issuecomment-555546793 that you were going to review the dataflow examples to figure out destroy behavior. Have you finished that?

callppatel commented:
> estroy behavior. Have you finished that?

No, I may need help on this as dataflow resource is created by custom code not in product list, could you please guide how it should look like, I mean how can we combine only detroy with custom go code 

Hope rest of the code apart from destroy is ok.

danawillow commented:
You can do it with `custom_delete`-  there's a few other resources that use it that you can reference at https://github.com/GoogleCloudPlatform/magic-modules/search?q=custom_delete&unscoped_q=custom_delete. And then you can model your go code off of what dataflow does.

I'll run the tests and re-review the rest once this part is done.

callppatel commented:
@danawillow - We dont have any use case of Terraform destroy, so I have added just skip destroy in terraform.yaml file like many others resource.  

custom_delete: templates/terraform/custom_delete/skip_delete.go.erb

modular-magician commented:
Hi!  I'm the modular magician, I work on Magic Modules.
I see that this PR has already had some downstream PRs generated. Any open downstreams are already updated to your most recent commit, 630109c7f0a301c552739238bb0802bce0b20531. 
## Pull request statuses
terraform-provider-google-beta already has an open PR.
terraform-google-conversion already has an open PR.
terraform-provider-google already has an open PR.
No diff detected in Ansible.
No diff detected in Inspec.
## New Pull Requests
I didn't open any new pull requests because of this PR.

danawillow commented:
I ran the tests again, and they're still failing:

```
------- Stdout: -------
=== RUN   TestAccMLEngineJob_mlJobTrainingExample
=== PAUSE TestAccMLEngineJob_mlJobTrainingExample
=== CONT  TestAccMLEngineJob_mlJobTrainingExample
--- FAIL: TestAccMLEngineJob_mlJobTrainingExample (0.00s)
    testing.go:635: Step 0 error: Argument or block definition required: On /opt/teamcity-agent/temp/buildTmp/tf-test496143597/main.tf line 2: An argument or block definition is required here.
FAIL
```

```
------- Stdout: -------
=== RUN   TestAccMLEngineJob_mlJobTensorExample
=== PAUSE TestAccMLEngineJob_mlJobTensorExample
=== CONT  TestAccMLEngineJob_mlJobTensorExample
--- FAIL: TestAccMLEngineJob_mlJobTensorExample (0.00s)
    testing.go:635: Step 0 error: Missing item separator: On /opt/teamcity-agent/temp/buildTmp/tf-test304234621/main.tf line 7: Expected a comma to mark the beginning of the next item.
FAIL
```

```
------- Stdout: -------
=== RUN   TestAccMLEngineJob_mlJobContainerExample
=== PAUSE TestAccMLEngineJob_mlJobContainerExample
=== CONT  TestAccMLEngineJob_mlJobContainerExample
--- FAIL: TestAccMLEngineJob_mlJobContainerExample (1.20s)
    testing.go:635: Step 0 error: errors during apply:
        
        Error: Error creating Job: googleapi: Error 400: Field: master_config.image_uri Error: Failed to read the container uri [gcr.io/test-proj-234209/mnist_pytorch_custom_container:mnist_pytorch_cpuj7gujjdqrb], Please make sure that the image exists and CloudML Engine service account has access to it.
        
          on /opt/teamcity-agent/temp/buildTmp/tf-test125803328/main.tf line 2:
          (source code not available)
        
        
FAIL
```

The first two look like basic test syntax errors, so they should be pretty easily fixable. For the container one, are you willing to make your test container publicly readable? If not, it might be worth replacing that test with something else, since I don't think we have the ability in Terraform right now to build and push a new container that we can use in the test.

callppatel commented:
> est syn
These TF test code are written in TF 11 and runs fine at my end. Let me check in TF12.
 
For container one. You might need to refer document https://cloud.google.com/ml-engine/docs/using-containers to set-up environment/gcp-project for this test.

danawillow commented:
Do you have a Dockerfile that we could use for this? Setting up an arbitrary container using instructions that are meant for someone with a use case in mind is outside of the scope of what can reasonably be expected as a code reviewer.

callppatel commented:
> Do you have a Dockerfile that we could use for this? Setting up an arbitrary container using instructions that are meant for someone with a use case in mind is outside of the scope of what can reasonably be expected as a code reviewer.

Docker file is located https://github.com/GoogleCloudPlatform/cloudml-samples/blob/master/pytorch/containers/quickstart/mnist/Dockerfile

Used the sample one which is documented [here](https://cloud.google.com/ml-engine/docs/custom-containers-training)

callppatel commented:
@danawillow - Could you please fix the error at your end and proceed with this pull request. This has been open for very long time. Our organization still use TF 11 so I dont see error, we are still in process of migrating to TF12. But you can make change at your end for test case to move ahead. Thanks for your support.

danawillow commented:
The tests run using the version of the Terraform SDK that is vendored into the provider repository, not the one installed on your machine. If you run the tests using the instructions at https://github.com/GoogleCloudPlatform/magic-modules#testing-your-changes, you should be able to get the same output that I do when I run it. If you're running the tests and they succeed, please paste the test output here, and then I'll take another look.

callppatel commented:
Our test is do terraform apply and terraform destory after build. I am attaching screen print of this. As of now, I have removed two test cases which require other input.

Thanks for your support !
[Testing Screen Prints.docx](https://github.com/GoogleCloudPlatform/magic-modules/files/3975430/Testing.Screen.Prints.docx)

callppatel commented:
screen print of container one test attached as well.
[Container Testing Screen Prints.docx](https://github.com/GoogleCloudPlatform/magic-modules/files/3975564/Container.Testing.Screen.Prints.docx)

callppatel commented:
@danawillow - Hope you are able to make progress on test. Looking fwd for your support to move this quickly this. Thank you !

callppatel commented:
This is not needed anymore, we are looking for other alternative than terraform for this.
